File: JavaSpring/src/main/java/com/magnetaryggdrasil/javaspring/service/TaskService.java
package com.magnetaryggdrasil.javaspring.service;

import com.magnetaryggdrasil.javaspring.model.Task;
import com.magnetaryggdrasil.javaspring.model.User;
import com.magnetaryggdrasil.javaspring.repository.TaskRepository;
import com.magnetaryggdrasil.javaspring.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class TaskService {

    private final TaskRepository taskRepository;
    private final UserRepository userRepository;

    @Autowired
    public TaskService(TaskRepository taskRepository, UserRepository userRepository) {
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
    }

    public Task createTask(Long userId, Task task) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found with id " + userId));
        task.setUser(user);
        return taskRepository.save(task);
    }

    public List<Task> getTasksByUserId(Long userId) {
        if (!userRepository.existsById(userId)) {
            throw new IllegalArgumentException("User not found with id " + userId);
        }
        return taskRepository.findByUserId(userId);
    }

    public Optional<Task> getTaskById(Long id) {
        return taskRepository.findById(id);
    }

    public Task updateTask(Long id, Task taskDetails) {
        return taskRepository.findById(id).map(task -> {
            if (taskDetails.getTitle() != null) {
                task.setTitle(taskDetails.getTitle());
            }
            if (taskDetails.getDescription() != null) {
                task.setDescription(taskDetails.getDescription());
            }
            // isDone is handled by specific endpoint, but let's allow it here too if needed, though requirements say "updates" and also "PATCH /done"
            if (taskDetails.getIsDone() != null) {
                task.setIsDone(taskDetails.getIsDone());
            }
            return taskRepository.save(task);
        }).orElseThrow(() -> new IllegalArgumentException("Task not found with id " + id));
    }

    public Task markTaskAsDone(Long id) {
        return taskRepository.findById(id).map(task -> {
            task.setIsDone(true);
            return taskRepository.save(task);
        }).orElseThrow(() -> new IllegalArgumentException("Task not found with id " + id));
    }

    public void deleteTask(Long id) {
        if (!taskRepository.existsById(id)) {
            throw new IllegalArgumentException("Task not found with id " + id);
        }
        taskRepository.deleteById(id);
    }
}

File: Cpp/src/services/TaskService.cc
#include "TaskService.h"

namespace services {

    TaskService::TaskService() {}

    std::optional<models::Task> TaskService::createTask(int userId, const std::string& title, const std::string& description) {
        if (!userRepo.exists(userId)) return std::nullopt;

        models::Task task;
        task.title = title;
        task.description = description;
        task.user_id = userId;
        task.is_done = false;

        int id = taskRepo.create(task);
        task.id = id;
        return task;
    }

    std::optional<models::Task> TaskService::getTaskById(int id) {
        return taskRepo.findById(id);
    }

    std::optional<std::vector<models::Task>> TaskService::getTasksByUserId(int userId) {
        if (!userRepo.exists(userId)) return std::nullopt;
        return taskRepo.findAllByUserId(userId);
    }

    std::optional<models::Task> TaskService::updateTask(int id, const std::string& title, const std::string& description) {
        auto task = taskRepo.findById(id);
        if (!task) return std::nullopt;

        if (!title.empty()) task->title = title;
        if (!description.empty()) task->description = description;

        taskRepo.update(*task);
        return task;
    }

    bool TaskService::markTaskAsDone(int id) {
        if (!taskRepo.exists(id)) return false;
        taskRepo.markAsDone(id);
        return true;
    }

    bool TaskService::deleteTask(int id) {
        if (!taskRepo.exists(id)) return false;
        taskRepo.deleteById(id);
        return true;
    }
}
